doc/developer/design/20230330_recursion_limit.md

```diff
@@ -0,0 +1,187 @@
+
+## Where to set the option?
+
+TLDR: I propose to put the hard limit (for use case 1.) in a system variable that is modifiable by users, and put the soft limit (for use case 2.) in a session variable.
+
+A system variable applies to all queries in the entire system. (Even if we make a dataflow pick up the current value only at dataflow creation, this can affect all queries if a restart suddenly happens.) Therefore, a system variable is not suitable for the 2. and 3. use cases. However, it is suitable for 1. (the hard limit).
+
+[Most system variables can't be modified by a user on her own, but only by support.](https://materialize.com/docs/sql/show/#system-variables) However, [there is a precedent](https://materializeinc.slack.com/archives/C02FWJ94HME/p1680209826137869?thread_ts=1680209764.706779&cid=C02FWJ94HME) for a system variable that is modifiable by users. As mentioned above, we would like the hard limit to be easily modifiable by users, and therefore if we decide to put the hard limit in a system variable, we should definitely make it modifiable by users.
+
+Session variables have the problem that a custom value will be lost across environmentd restarts, meaning that it is impossible to install a view where the limit is different from the default, and have the custom limit stick between restarts. (We could conceivably add the value to the catalog along with the query's SQL definition, but this is probably not worth the trouble.) Therefore, a session variable wouldn't be suitable for the 1. (the hard limit) and 3. use cases, because for these one has to raise the limit for a production view in a manner that survives restarts. However, it is suitable for the 2. use case (debugging): one can temporarily change the limit, and run the query that is being debugged. Note that setting the limit shouldn't change the limit in dataflows that are already running, to avoid messing those up during a debugging session of a new query.
+
+To sum up, the variables approach seems to be suitable for both the hard limit (use case 1.) and soft limit (use case 2.). I propose to implement these very soon. I already have the rendering code ready to go, and adding the variables and tests should be doable in just a few hours.
+
+Another approach would be to put the options into the queries SQL definitions, which I discuss in [future work](#future-work).
+
+## Workaround: Manually setting a limit using existing SQL building blocks
+
+[Users can manually write SQL that counts the number of iterations, and converts reaching the iteration count limit into a fixpoint.](https://github.com/MaterializeInc/materialize/issues/18362#issuecomment-1481605865)
+Note that this is useful only for use case 2. (debugging). It's not useful for 1., because hitting the limit happens unexpectedly there. It might be useful for 3. later, when we have broadcast joins.
```

Comment:
If I understood https://github.com/MaterializeInc/materialize/issues/18362#issuecomment-1481605865 correctly, it includes a bad cross join if we don't have broadcast joins. This is kind of ok for use case 2. (because the data size can be kept small for debugging), but not ok for use case 3.

Reply:
Ah, yes. OK, I think I understand better the distinction between (2) and (3) now.

2. Debugging intermediate results of queries with toy data during development in order to track down and fix correctness or non-termination issues.
3. Debugging intermediate results of queries with bigger data sets (presumably from a production or end-to-end testing environments) in order to track down and fix correctness or non-termination issues that were not caught during development but are somehow manifesting on the bigger data set.

So the difference is not so much of the intent or the type of information you want to get out (for example: all intermediate results vs a specific intermediate result), vs. the amount of data that you are dealing with. So it seems that if we have a solution that works equally well with different data / delta sizes we are fine with using the same solution and UX / ergonomics for both use cases. Is that correct?

Reply:
Well, 3. would be not just for debugging, but maybe some algorithm requires a fixed number of steps. E.g., maybe you want to do hyperparameter optimization and you have a specific set of, say, 20 parameter settings to try.

Reply:
I think if we have a per-WMR limit that is
- expressed in SQL, so works across restarts,
- works with all data sizes,
then it would be usable for both 2. and 3.